### PR TITLE
Port to GTK 4

### PR DESCRIPTION
**Pull request by danirabbit**
Closes #412

### CODE REVIEW

File: src/Dialogs/ProgressDialog.vala [hunk 1]
```diff
@@ -62,17 +63,16 @@ public class Keyboard.InputMethodPage.ProgressDialog : Granite.Dialog {
 
         var grid = new Gtk.Grid () {
             column_spacing = 12,
-            margin = 6,
+            margin_start = 12,
+            margin_end = 12,
```
Comment (danirabbit) [inline src/Dialogs/ProgressDialog.vala:67]: Double check this is right

File: src/InputSources/SourceSettings.vala [hunk 2]
```diff
@@ -48,7 +48,7 @@ class Keyboard.SourceSettings : Object {
         }
     }
 
-    private GLib.List<InputSource> input_sources;
+    public unowned GLib.List<InputSource> input_sources { get; private set; }
```
Comment (danirabbit) [inline src/InputSources/SourceSettings.vala:51]: Maybe better to add get length method?

File: src/Shortcuts/Shortcut.vala [hunk 3]
```diff
@@ -69,22 +69,10 @@ public class Keyboard.Shortcuts.Shortcut : GLib.Object {
             tmp += _("Ctrl") + SEPARATOR;
         }
 
-        if ((modifiers & Gdk.ModifierType.MOD1_MASK) > 0) {
+        if ((modifiers & Gdk.ModifierType.ALT_MASK) > 0) {
             tmp += _("Alt") + SEPARATOR;
         }
 
-        if ((modifiers & Gdk.ModifierType.MOD2_MASK) > 0) {
```
Comment (danirabbit) [inline src/Shortcuts/Shortcut.vala:76]: These don't exist in Gdk anymore :shrug:

File: src/Dialogs/ConflictDialog.vala [hunk 4]
```diff
@@ -36,7 +36,7 @@ public class ConflictDialog : Granite.MessageDialog {
         add_button (_("Cancel"), Gtk.ResponseType.CANCEL);
 
         var reassign_button = add_button (_("Reassign"), Gtk.ResponseType.ACCEPT);
-        reassign_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+        reassign_button.get_style_context ().add_class (Granite.STYLE_CLASS_DESTRUCTIVE_ACTION);
```
Comment (jeremypw) [inline src/Dialogs/ConflictDialog.vala:39]: Use `.add_css_class ()` ?

File: src/Dialogs/ProgressDialog.vala [hunk 5]
```diff
@@ -38,7 +39,7 @@ public class Keyboard.InputMethodPage.ProgressDialog : Granite.Dialog {
             wrap = true,
             xalign = 0
         };
-        primary_label.get_style_context ().add_class (Granite.STYLE_CLASS_PRIMARY_LABEL);
+        primary_label.get_style_context ().add_class (Granite.STYLE_CLASS_TITLE_LABEL);
```
Comment (jeremypw) [inline src/Dialogs/ProgressDialog.vala:42]: Use `add_css_class ()` ?

File: src/Shortcuts/Shortcut.vala (hunk 3 above)
Comment (jeremypw) [reply src/Shortcuts/Shortcut.vala:76]: What about META_MASK, which exists in Gdk4?

File: src/Widgets/InputMethod/AddEnginesPopover.vala [hunk 6]
```diff
@@ -54,7 +54,7 @@ public class Keyboard.InputMethodPage.AddEnginesPopover : Gtk.Popover {
         var cancel_button = new Gtk.Button.with_label (_("Cancel"));
 
         var add_button = new Gtk.Button.with_label (_("Add Engine"));
-        add_button.get_style_context ().add_class (Gtk.STYLE_CLASS_SUGGESTED_ACTION);
+        add_button.get_style_context ().add_class (Granite.STYLE_CLASS_SUGGESTED_ACTION);
```
Comment (jeremypw) [inline src/Widgets/InputMethod/AddEnginesPopover.vala:57]: `add_css_class ()`?

**Comment by jeremypw** [comment]
Adding and removing layouts and input engines seems to work OK except that adding an input engine (Japanese Anthy)  and immediately removing it invalidates the installed keyboard layout (English (UK)) - however master has the same issue.  Shortcuts seems to work OK except cannot use `Fn` as modifier for custom shortcuts.  I cannot check whether `master` has the same problem as Custom Shortcuts has stopped working there.   

The cursor blinking scale seems to operate the unexpected way?  Increase the value slows the speed of blinking even thought the header is "Speed" - but this is the same as master.  The duration scale did not have a noticeable effect - same as master.  Maybe reconsider the wording of the labels e.g. "Interval on" and "Interval off"?

Sound setings hyperlink not working - presumably because not ported yet.

**Review by jeremypw — CHANGES_REQUESTED** [review]
In "Behaviour" clicking on the Hyperlink to Onboard settings works but also throws an error dialog "Could not show link - launched failed" - this issue is not present in master.

**Comment by danirabbit** [comment]
@jeremypw thanks for the review! For any issues that also exist in master could you open issue reports please? This is just a port and doesn't really seek to solve existing issues.

Sound settings link is working here. Do you have the `io.elementary.settings.sound` installed?

Fixed the linkbutton issue!

**Comment by jeremypw** [comment]
> o you have the io.elementary.settings.sound installed?

Ah, no.  That would explain it!

**Review by jeremypw — APPROVED** [review]

**Pull request merged**